api-server/src/server/boot/challenge.js

```diff
@@ -429,35 +502,178 @@ async function backendChallengeCompleted(req, res, next) {
+        res.status(500);
+        throw new Error(
+          `An error occurred trying to get the exam from the database.`
+        );
+      }
+
+      // This is cause there was struggles validating the exam directly from the db/loopback
+      const examJson = JSON.parse(JSON.stringify(examFromDb));
+
+      const validExamFromDbSchema = validateExamFromDbSchema(examJson);
+
+      if (validExamFromDbSchema.error) {
+        res.status(500);
+        log(validExamFromDbSchema.error);
+        throw new Error(
+          `An error occurred validating the exam information from the database.`
+        );
+      }
+
+      const { prerequisites, numberOfQuestionsInExam, title } = examJson;
+
+      // Validate User has completed prerequisite challenges
+      prerequisites?.forEach(prerequisite => {
+        const prerequisiteCompleted = user.completedChallenges.find(
+          challenge => challenge.id === prerequisite.id
+        );
+
+        if (!prerequisiteCompleted) {
+          res.status(403);
+          throw new Error(
+            `You have not completed the required challenges to start the '${title}'.`
+          );
+        }
+      });
+
+      const randomizedExam = generateRandomExam(examJson);
+
+      const validGeneratedExamSchema = validateGeneratedExamSchema(
+        randomizedExam,
+        numberOfQuestionsInExam
+      );
+
+      if (validGeneratedExamSchema.error) {
+        res.status(500);
+        log(validGeneratedExamSchema.error);
+        throw new Error(`An error occurred trying to randomize the exam.`);
+      }
+
+      return res.send({ generatedExam: randomizedExam });
+    } catch (err) {
+      log(err);
+      return res.send({ error: err.message });
     }
+  };
+}
 
-    return res.json({
-      alreadyCompleted,
-      points: alreadyCompleted ? user.points : user.points + 1,
-      completedDate: completedChallenge.completedDate
-    });
-  });
+function createExamChallengeCompleted(app) {
+  const { Exam } = app.models;
+
+  return async function examChallengeCompleted(req, res, next) {
+    const { body = {}, user } = req;
+
+    try {
+      await user.getCompletedChallenges$().toPromise();
+    } catch (e) {
+      return next(e);
+    }
+
+    const { userCompletedExam = [], id } = body;
+
+    try {
+      const examFromDb = await Exam.findById(id);
+      if (!examFromDb) {
+        res.status(500);
+        throw new Error(
+          `An error occurred tryng to get the exam from the database.`
+        );
+      }
+
+      // This is cause there was struggles validating the exam directly from the db/loopback
+      const examJson = JSON.parse(JSON.stringify(examFromDb));
+
+      const validExamFromDbSchema = validateExamFromDbSchema(examJson);
+      if (validExamFromDbSchema.error) {
+        res.status(500);
+        log(validExamFromDbSchema.error);
+        throw new Error(
+          `An error occurred validating the exam information from the database.`
+        );
+      }
```

Comment:
I say we leave it for now. Would the better route be to then validate everything in the `generateExam` function? I don't do any validation in there cause I do it here.

Reply:
With the new api this will be largely irrelevant since _most_ of what this does will be handled by Prisma + TypeScript.

That said, the most performant approach is to validate as the data is added to the database, but validating in the api does provide an extra layer of protection.

I'm not sure, really, but I think it's something to keep in mind.

Reply:
It is being validate before it goes into the database - I feel like we should validate it when it comes out, too. This ensures (hopefully) that everything will go smoothly when we run the `generateExam` function with the data - or, like I said, maybe change it to just have some error handling within that function instead - or have both I guess.

Reply:
Sure. I'm happy with this approach.